libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+              const hash: crypto.Hash = crypto.createHash('sha1');
+              for (const operation of set) {
+                const { associatedPhase: phase, associatedProject: project } = operation;
+                if (project && phase) {
+                  hash.update(project.projectRelativeFolder);
+                  hash.update(RushConstants.hashDelimiter);
+                  hash.update(phase.name);
+                  hash.update(RushConstants.hashDelimiter);
+                }
+              }
+              const cobuildClusterId: string = hash.digest('hex');
+
+              // Assign same cluster id to all operations in the same cluster.
+              for (const operation of set) {
+                const { runner } = operation;
+                if (runner instanceof ShellOperationRunner) {
+                  const buildCacheContext: IOperationBuildCacheContext =
+                    this._getBuildCacheContextByRunnerOrThrow(runner);
+                  buildCacheContext.cobuildClusterId = cobuildClusterId;
+                }
+              }
+            }
+          }
+        }
+
+        return operations;
+      }
+    );
+
+    hooks.afterExecuteOperation.tapPromise(
+      PLUGIN_NAME,
+      async (runnerContext: IOperationRunnerContext): Promise<void> => {
+        const { runner, status, consumers, changedProjectsOnly } = runnerContext;
+        const buildCacheContext: IOperationBuildCacheContext | undefined =
+          this._getBuildCacheContextByRunner(runner);
+
+        let blockCacheWrite: boolean = !buildCacheContext?.isCacheWriteAllowed;
+        let blockSkip: boolean = !buildCacheContext?.isSkipAllowed;
+
+        switch (status) {
+          case OperationStatus.Skipped: {
+            // Skipping means cannot guarantee integrity, so prevent cache writes in dependents.
+            blockCacheWrite = true;
+            break;
+          }
+
+          case OperationStatus.SuccessWithWarning:
+          case OperationStatus.Success: {
+            // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+            blockSkip ||= !changedProjectsOnly;
+            break;
+          }
+        }
+
+        // Apply status changes to direct dependents
+        for (const item of consumers) {
+          const itemRunnerBuildCacheContext: IOperationBuildCacheContext | undefined =
+            this._getBuildCacheContextByRunner(item.runner);
+          if (itemRunnerBuildCacheContext) {
+            if (blockCacheWrite) {
+              itemRunnerBuildCacheContext.isCacheWriteAllowed = false;
+            }
+            if (blockSkip) {
+              itemRunnerBuildCacheContext.isSkipAllowed = false;
+            }
+          }
+        }
+      }
+    );
+
+    hooks.afterExecuteOperations.tapPromise(PLUGIN_NAME, async () => {
+      for (const { buildCacheProjectLogWritable } of this._buildCacheContextByOperationRunner.values()) {
+        buildCacheProjectLogWritable?.close();
```

Comment:
Is this here to allow for cache writes to happen asynchronously while consumers are executing?

Reply:
Originally YES.
Now, I have refactored this portion of the logic based on your suggestion, ensuring that the lifetime of the `ProjectLogWritable` is appropriately wrapped between the `beforeExecute` and `afterExecute` hooks.